README.md

# Xsolla Login Javascript SDK

Currently SDK supports following types of authorization:
* via login/password
* via social networks

On this page you can see:
<!-- TOC depthFrom:1 depthTo:6 withLinks:1 updateOnSave:1 orderedList:0 -->

- [Integrating of the Widget](#integrating-of-the-widget)
	- [Step 1: Connecting SDK](#step-1-connecting-sdk)
	- [Step 2: Initializing Widget](#step-2-initializing-widget)
	- [Step 3: Choosing the Widget Placing](#step-3-choosing-the-widget-placing)
		- [Fullscreen Mode](#fullscreen-mode)
		- [Block on the Page](#block-on-the-page)
- [Usage](#usage)
	- [Customizing Widget Style](#customizing-widget-style)
	- [Tracking Widget Events](#tracking-widget-events)
		- [Loading Widget](#loading-widget)
		- [Closing Widget](#closing-widget)
		- [Sending the Registration Confirmation Email](#sending-the-registration-confirmation-email)

<!-- /TOC -->

See [documentation](https://developers.xsolla.com/doc/login/) to find more.

## Integrating of the Widget

This library allows you to quickly integrate Xsolla Login Widget with your website in following steps:
* [Connecting SDK](#step-1-connecting-sdk)
* [Initializing Widget](#step-2-initializing-widget)
* [Choosing the Widget Placing](#step-3-choosing-the-widget-placing)

### Step 1: Connecting SDK

Connect Xsolla Login Javascript SDK using one of the following methods:
* If the project uses [Bower](http://bower.io), launch the console and run
```shell script
bower install xsolla-login-js-sdk
```
* If the package is not connected, add the following code to the `<head>` tag of the HTML-page where the widget will be placed:
```html
<script src="https://cdn.xsolla.net/xsolla-login-widget/sdk/2.2.6/xl.min.js"></script>
```

### Step 2: Initializing Widget

Add the widget initialization code to the `<body>` tag.
```html
<script type="text/javascript">
XL.init({
  projectId: '[Login ID]',
  callbackUrl: '[callbackUrl]',
  locale: 'en_US'
});
</script>
```

--------------------------------------------------------------------------------

  **projectId**  
  &nbsp;&nbsp;&nbsp;&nbsp; `string`  
  Login ID from Publisher Account. **Required**.  

 ------------------------------------------------------------------
  **callbackUrl**  
  &nbsp;&nbsp;&nbsp;&nbsp; `string`  
  URL to redirect the user to after registration/authentication/password reset. Must be identical to one of the **Callback URL** specified in Publisher Account in Login settings. **Required** if there are several Callback URLs.

  You can pass several URL of the local server to make the widget available to the local build. For example `https://localhost:9000`.

 ------------------------------------------------------------------
  **locale**  
  &nbsp;&nbsp;&nbsp;&nbsp; `string`  
  Region in the `<language code>_<country code>` format, where:
  * *language code*: language code in the [ISO 639-1](https://en.wikipedia.org/wiki/ISO_639-1) format;
  * *country code*: country/region code in the [ISO 3166-1 alpha-2](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2) format.

  The language of the widget interface. Also used for sorting social networks by the frequency of use in the chosen region.

 ------------------------------------------------------------------
  **fields**  
  &nbsp;&nbsp;&nbsp;&nbsp; `array of strings`  
  List of parameters required for successful user registration/authentication. Example: '**email, promo_email_agreement**'. The user will be asked to specify these parameters in the corresponding form.

 ------------------------------------------------------------------
  **theme**  
  &nbsp;&nbsp;&nbsp;&nbsp; `string`  
  URL with the widget styles file. If the value is empty, styles uploaded to Publisher Account are used.  

 ------------------------------------------------------------------
  **popupBackgroundColor**  
  &nbsp;&nbsp;&nbsp;&nbsp; `string`  
  Widget background color in the **fullscreen** mode. The value can be in any of the CSS color formats. Default is rgba(50, 150, 150, 0.1).

 ------------------------------------------------------------------
  **iframeZIndex**  
  &nbsp;&nbsp;&nbsp;&nbsp; `string`  
  The stack order of the widget in the **fullscreen** mode. Default is 1000000.  

--------------------------------------------------------------------------------  

### Step 3: Choosing the Widget Placing

Choose the widget placing on the website start page:
* [Fullscreen Mode](#fullscreen-mode)
* [Block on the Page](#block-on-the-page)

#### Fullscreen Mode

Add the button with the `on-click` event and the `XL.show()` function to the site.
```html
<button onclick="XL.show()">Fullscreen widget</button>
```

The fullscreen mode is closed by clicking outside the widget area.

#### Block on the Page

Add the block with the widget to the `<body>` tag and set the block ID.
```html
<div id="xl_auth"></div>
```

Add the code and pass the parameter values as described below.
```html
<script type="text/javascript">
var element_id = 'xl_auth';
var options = {
  width: 400,
  height: 550,
  route: XL.ROUTES.REGISTRATION
};
XL.AuthWidget(element_id, options);
</script>
```

 --------------------------------------------------------------------------------

   **element_id**  
   &nbsp;&nbsp;&nbsp;&nbsp; `string`  
   ID of the block containing the Login Widget. **Required**.  

  ------------------------------------------------------------------
   **options**  
   &nbsp;&nbsp;&nbsp;&nbsp; `object`  
   Login Widget block settings. The object consists of the parameters listed below.

  ------------------------------------------------------------------
   **width**  
   &nbsp;&nbsp;&nbsp;&nbsp; `number`  
   Block width in pixels. Default is 400.

  ------------------------------------------------------------------
   **height**  
   &nbsp;&nbsp;&nbsp;&nbsp; `number`  
   Block height in pixels. Default is 550.

  ------------------------------------------------------------------
   **route**  
   &nbsp;&nbsp;&nbsp;&nbsp; `string`  
   Widget start page. Can be:
   * `XL.ROUTES.LOGIN` — a login page via username/password. Used by default.
   * `XL.ROUTES.REGISTRATION` — a registration page.
   * `XL.ROUTES.RECOVER_PASSWORD` — a password reset page.
   * `XL.ROUTES.ALL_SOCIALS` — a page with a list of social networks available for user authentication.

 --------------------------------------------------------------------------------  

## Usage

If you have already integrated Login, you can also try additional features of Xsolla Login Widget:
* [Customizing Widget Style](#customizing-widget-style)
* [Tracking Widget Events](#tracking-widget-events)

### Customizing Widget Style

By default, the widget looks like this:  
<img src="https://cdn3.xsolla.com/img/misc/images/08ef59d6c3bf6b3c133854cc579423fd.png" width="75%" title="Widget by default">

You can customize the widget style by changing the following characteristics of its elements:
* Location on the screen
* Theme
* Size
* Background  

Customization is based on [CSS](https://en.wikipedia.org/wiki/Cascading_Style_Sheets).

To customize:
1. Go to [Publisher Account](https://publisher.xsolla.com/) > your Login project > **Customization** > **Widget style**, download the archive with styles and unpack it.
2. Download and install [Node.js](https://nodejs.org/en/). The version of Node.js must not be later than 10.x.x.
3. Run `npm i` from the unpacked styles folder.
4. Edit *app/styles/themes/\_default.scss*.
5. Run `npm run styles` from the unpacked styles folder.
6. Upload the file with a name resembling *app.default.css* from the **dist** folder to your Publisher Account.  

For further widget UI customization, repeat steps 4–6.

---
**NOTE**

After publication, the widget will be changed for all the projects it was connected to.

---  

### Tracking Widget Events

You can collect widget statistics on the following events:
* [Loading Widget](#loading-widget)
* [Closing Widget](#closing-widget)
* [Sending the Registration Confirmation Email](#sending-the-registration-confirmation-email)

To start tracking the event, initialize and process the action as described below.

#### Loading Widget

```js
XL.on(XL.eventTypes.LOAD, function () {
    console.log('the widget has been loaded');
});
```

#### Closing Widget

```js
XL.on(XL.eventTypes.CLOSE, function () {
    console.log('user has closed the widget');
});
```

#### Sending the Registration Confirmation Email

```js
XL.on(XL.eventTypes.REGISTRATION_REQUEST, function () {
    console.log('registration form has been sent');
});
```

CHANGELOG.md
# Changelog

## 2.2.6 - 2020-02-12

- [chore]: Updated docs

## 2.2.5 - 2020-02-12

- [chore]: Updated docs

## 2.2.4 - 2019-10-08

- [chore]: Bump version

## 2.2.3 - 2019-07-19

- [feat]: Implemented OAuth 2.0 authentication 
  
## 2.1.2 - 2019-07-04

- [fix] Last custom theme is used instead of the default now.

## 2.1.1 - 2019-06-18

- [fix] Updated dependencies

## 2.1.0 - 2019-06-11

- [feat] Added compact parameter
- [feat] Added routes for only social and only username/pass login

## 2.0.0 - 2019-06-07

- [feat] Added widgetVersion parameter
- [feat] Added externalWindow parameter
- [feat] Added registration event

## 1.4.1 - 2019-04-10

- [fix] Fix fullscreen popup

## 1.3.2 - 2018-10-22

- [build] Updated dependencies

## 1.3.1 - 2018-08-01

- [feature] Added parameter for setting Widget theme
- [feature] Added parameter for opening Widget at specific route

## 1.1.1 - 2017-02-13

- [fix] Compatibility with IE 11

## 1.1.0 - 2017-02-09

- [feature] Event handler after widget was loaded
- [feature] Event handler after close button was clicked


test/xlapiSpec.js
'use strict';
var XLApi = require('../src/xlapi.js');


describe("Xsolla Login API", function() {
    var xlApi = new XLApi(123);

    it("says hello", function(done) {
        xlApi.getSocialsURLs(function (e) {

        }, function (e) {
            expect(e).toBe('Network error');
            done();
        });
    });
});


tests/main.test.js
const main = require('../src/main');
const ParseUrl = require('url-parse');

const TEST_CONFIG = {
    projectId: '40db2ea4-5d42-11e6-a3ff-005056a0e04a',
    redirectUrl: 'https://publisher.xsolla.com',
    callbackUrl: 'https://test.xsolla.com',
    loginUrl: 'https://test.xsolla.com',
    popupBackgroundColor: 'rgba(102, 255, 51, 1)',
    iframeZIndex: '0',
    theme: 'testTheme.css',
    locale: 'en_US',
    fields: 'username,email',
    externalWindow: 'true',
    route: 'XL.ROUTES.ALL_SOCIALS',
    widgetVersion: '4.0.1',
    widgetBaseUrl: 'https://base-widget-test.com/',
    compact: 'true',
    onlyWidgets: 'true',
    state: 'ABCD1234',
    clientId: '1',
    redirectUri: 'http://example.com',
    responseType: 'code'
};

beforeEach(() => {
    main.init({
        projectId: TEST_CONFIG.projectId,
        redirectUrl: TEST_CONFIG.redirectUrl,
        callbackUrl: TEST_CONFIG.callbackUrl,
        popupBackgroundColor: TEST_CONFIG.popupBackgroundColor,
        iframeZIndex: TEST_CONFIG.iframeZIndex,
        theme: TEST_CONFIG.theme,
        locale: TEST_CONFIG.locale,
        fields: TEST_CONFIG.fields,
        externalWindow: TEST_CONFIG.externalWindow,
        widgetVersion: TEST_CONFIG.widgetVersion,
        compact: TEST_CONFIG.compact,
        onlyWidgets: TEST_CONFIG.onlyWidgets,
        state: TEST_CONFIG.state,
        clientId: TEST_CONFIG.clientId,
        redirectUri: TEST_CONFIG.redirectUri,
        responseType: TEST_CONFIG.responseType,
    });
});

describe('Widget config params', function () {
    it('contains popupBackgroundColor param', () => {
        expect(main.config.popupBackgroundColor).toBe(TEST_CONFIG.popupBackgroundColor);
    });

    it('contains iframeZIndex param', () => {
        expect(main.config.iframeZIndex).toBe(TEST_CONFIG.iframeZIndex);
    });
});

describe('Get iframe source', function () {
    it('returns default value', () => {
        const defaultUrl = ParseUrl(main.getLink(), true);

        expect(defaultUrl.origin).toBe('https://xl-widget.xsolla.com');
    });

    it('changes widgetBaseUrl param', () => {
        const defaultUrl = ParseUrl(main.getLink({ widgetBaseUrl: TEST_CONFIG.widgetBaseUrl }), true);

        expect(defaultUrl.origin + '/').toBe(TEST_CONFIG.widgetBaseUrl);
    });

    it('changes route param', () => {
        const defaultUrl = ParseUrl(main.getLink({ route: TEST_CONFIG.route }), true);

        expect(defaultUrl.pathname.substring(1)).toBe(TEST_CONFIG.route);
    });

    it('contains projectId param', () => {
        const defaultUrl = ParseUrl(main.getLink(), true);

        expect(decodeURIComponent(defaultUrl.query.projectId)).toBe(TEST_CONFIG.projectId);
    });

    it('contains locale param', () => {
        const defaultUrl = ParseUrl(main.getLink(), true);

        expect(decodeURIComponent(defaultUrl.query.locale)).toBe(TEST_CONFIG.locale);
    });

    it('contains fields param', () => {
        const defaultUrl = ParseUrl(main.getLink(), true);

        expect(decodeURIComponent(defaultUrl.query.fields)).toBe(TEST_CONFIG.fields);
    });

    it('contains redirectUrl param', () => {
        const defaultUrl = ParseUrl(main.getLink(), true);

        expect(decodeURIComponent(defaultUrl.query.redirectUrl)).toBe(TEST_CONFIG.redirectUrl);
    });

    it('contains loginUrl param', () => {
        const defaultUrl = ParseUrl(main.getLink(), true);

        expect(decodeURIComponent(defaultUrl.query.login_url)).toBe(TEST_CONFIG.loginUrl);
    });

    it('contains theme param', () => {
        const defaultUrl = ParseUrl(main.getLink(), true);

        expect(decodeURIComponent(defaultUrl.query.theme)).toBe(TEST_CONFIG.theme);
    });

    it('contains externalWindow param', () => {
        const defaultUrl = ParseUrl(main.getLink(), true);

        expect(decodeURIComponent(defaultUrl.query.external_window)).toBe(TEST_CONFIG.externalWindow);
    });

    it('contains widgetVersion param', () => {
        const defaultUrl = ParseUrl(main.getLink(), true);

        expect(decodeURIComponent(defaultUrl.query.version)).toBe(TEST_CONFIG.widgetVersion);
    });

    it('contains compact param', () => {
        const defaultUrl = ParseUrl(main.getLink(), true);

        expect(decodeURIComponent(defaultUrl.query.compact)).toBe(TEST_CONFIG.compact);
    });

    it('contains state param', () => {
        const defaultUrl = ParseUrl(main.getLink(), true);

        expect(decodeURIComponent(defaultUrl.query.state)).toBe(TEST_CONFIG.state);
    });

    it('should randomize state by default', () => {
        main.init({
            projectId: 'test',
            clientId: '1'
        });
        const url = ParseUrl(main.getLink(), true);

        expect(decodeURI(url.query.state).length).toBeGreaterThan(8);
    });

    it('contains client_id param', () => {
        const defaultUrl = ParseUrl(main.getLink(), true);

        expect(decodeURIComponent(defaultUrl.query.client_id)).toBe(TEST_CONFIG.clientId);
    });

    it('contains redirect_uri param', () => {
        const defaultUrl = ParseUrl(main.getLink(), true);

        expect(decodeURIComponent(defaultUrl.query.redirect_uri)).toBe(TEST_CONFIG.redirectUri);
    });

    it('contains response_type param', () => {
        const defaultUrl = ParseUrl(main.getLink(), true);

        expect(decodeURIComponent(defaultUrl.query.response_type)).toBe(TEST_CONFIG.responseType);
    });

    it('should not contain project_id', () => {
        main.init({
            clientId: 'YOUR CLIENT ID',
            redirectUri: 'http://localhost:8888/callback',
            projectId: 'YOUR PROJECT ID'
        });

        const defaultUrl = ParseUrl(main.getLink(), true);

        expect(decodeURIComponent(defaultUrl.query.project_id)).toBe("undefined");
    });

    it('should not contain compact', () => {
        main.init({
            clientId: 'YOUR CLIENT ID',
            redirectUri: 'http://localhost:8888/callback',
            projectId: 'YOUR PROJECT ID'
        });
        const defaultUrl = ParseUrl(main.getLink(), true);

        expect(decodeURIComponent(defaultUrl.query.compact)).toBe("undefined");
    });

    it('response_type should by "code" by default', () => {
        main.init({
            projectId: 'test',
            clientId: '1'
        });
        const url = ParseUrl(main.getLink(), true);

        expect(decodeURIComponent(url.query.response_type)).toBe('code');
    });
});

gulpfile.js
/**
 * Created by a.korotaev on 14.06.16.
 */

var gulp = require('gulp');
var gutil = require('gulp-util');
var uglify = require('gulp-uglify');
var rename = require('gulp-rename');
var sourcemaps = require('gulp-sourcemaps');
var source = require('vinyl-source-stream');
var buffer = require('vinyl-buffer');
var browserSync = require('browser-sync');
var browserify = require('browserify');
var watchify = require('watchify');
var gulpif = require('gulp-if');
var commonShake = require('common-shakeify');
var strip = require('gulp-strip-comments');

function setupBrowserify(watch) {
    var bundleOptions = {
        cache: {},
        packageCache: {},
        paths: ['./src'],
        standalone: 'XL',
        fullPaths: false,
        debug: true,
        transform: [["babelify", { "presets": ["es2015"], "plugins": ["transform-object-assign"] }]]
    };
    var bundler = browserify('./src/main.js', bundleOptions).plugin(commonShake);
    bundler.require('./src/main.js', {entry: true, expose: 'main'});

    if (watch) {
        bundler = watchify(bundler);
        bundler.on('update', function () {  // on any dep update, runs the bundler
            runBundle(bundler, watch);
        });
    }

    runBundle(bundler, watch);
}

function runBundle(bundler, watch) {
    return bundler.bundle()
    // log errors if they happen
        .on('error', gutil.log.bind(gutil, 'Browserify Error'))
        .pipe(source('xl.js'))
        .pipe(buffer())
        .pipe(strip({
            ignore: /\/\/#/g
        }))
        .pipe(gulp.dest('./dist'))
        .pipe(buffer())
        .pipe(rename('xl.min.js'))
        .pipe(sourcemaps.init({loadMaps: true})) // loads map from browserify file
        .pipe(uglify())
        .pipe(sourcemaps.write('./', {includeContent: true, sourceRoot: '.', debug: false})) // writes .map file
        .pipe(gulp.dest('./dist'))
        .pipe(gulpif(watch, browserSync.reload({stream: true, once: true})));
}

gulp.task('build', function (done) {
    setupBrowserify(false);
    done();
});

gulp.task('browser-sync', function () {
    browserSync({
        startPath: 'example/widget.html',
        server: {
            baseDir: './'
        },
        port: 9000,
        ghostMode: false
    });
});

gulp.task('serve', gulp.series(['browser-sync'], function () {
    setupBrowserify(true);

    gulp.watch(['example/*.html']).on('change', browserSync.reload); //all the other files are managed by watchify
}));

src/xlapi.js
/**
 * Created by a.korotaev on 24.06.16.
 */
/**
 * Impelements Xsolla Login Api
 * @param projectId - project's unique identifier
 * @param baseUrl - api endpoint
 * @constructor
 */

var XLApi = function (projectId, baseUrl) {
    var self = this;
    this.baseUrl = baseUrl || '//login.xsolla.com/api/';

    this.projectId = projectId;

    this.makeApiCall = function (params, success, error) {
        var r = new XMLHttpRequest();
        r.withCredentials = true;
        r.open(params.method, self.baseUrl + params.endpoint, true);
        r.onreadystatechange = function () {
            if (r.readyState == 4) {
                if (r.status == 200) {
                    success(JSON.parse(r.responseText));
                } else {
                    if (r.responseText) {
                        error(JSON.parse(r.responseText));
                    } else {
                        error({error: {message: 'Networking error', code: r.status}});
                    }
                }
            }
        };
        if (params.method == 'POST') {
            r.setRequestHeader("Content-Type", "application/json;charset=UTF-8");
            r.send(params.postBody);
        } else if (params.method == 'GET') {
            r.send(params.getArguments);
        }
    };
};
/**
 * Get all avialable social methods auth url
 * @param success - success callback
 * @param error - error callback
 * @param getArguments - additional params to send with request
 */
XLApi.prototype.getSocialsURLs = function (success, error, getArguments) {
    var str = "";
    for (var key in getArguments) {
        if (str != "") {
            str += "&";
        }
        str += key + "=" + encodeURIComponent(getArguments[key]);
    }

    return this.makeApiCall({method: 'GET', endpoint: 'social/login_urls?' + str, getArguments: null}, success, error);
};

XLApi.prototype.loginPassAuth = function (login, pass, rememberMe, redirectUrl, success, error) {
    var body = {
        username: login,
        password: pass,
        remember_me: rememberMe
    };
    return this.makeApiCall({method: 'POST', endpoint: 'proxy/login?projectId='+this.projectId + '&redirect_url=' + encodeURIComponent(redirectUrl), postBody: JSON.stringify(body)}, success, error);
};

XLApi.prototype.smsAuth = function (phoneNumber, success, error) {
    return this.makeApiCall({method: 'GET', endpoint: 'sms', getArguments: 'phoneNumber=' + phoneNumber}, success, error);
};

module.exports = XLApi;


karma.conf.js
'use strict';
var istanbul = require('browserify-istanbul');

module.exports = function(karma) {
    karma.set({

        frameworks: [ 'browserify' ,  'jasmine'],

        files: [
            'src/**/*.js',
            // 'src/**/xlapi.js',
            // 'src/**/main.js',
            // 'dist/xl.js',
            'test/**/*Spec.js'
        ],

        reporters: [ 'coverage', 'dots', 'progress'],

        preprocessors: {
            'test/**/*Spec.js': [ 'browserify' ],
            'src/**/*.js': ['browserify', 'coverage' ]
            // 'src/**/main.js': ['browserify']
        },

        browsers: [ 'PhantomJS' ],

        logLevel: 'LOG_DEBUG',

        singleRun: true,
        autoWatch: false,

        // browserify configuration
        browserify: {
            debug: true,
            transform: [ 'brfs', 'browserify-shim' , istanbul({
                ignore: ['**/node_modules/**', '**/test/**']
            })]
        },
        coverageReporter: {
            reporters: [
                {type: 'html'},
                {type: 'cobertura'}
            ]
        }
    });
};

src/main.js
/**
 * Created by a.korotaev on 24.06.16.
 */
const toSnakeCase = require('to-snake-case');

require('./supports');
const version = require('../package.json').version;

import XLApi from './xlapi';
/**
 * Create an `Auth0` instance with `options`
 *
 * @class XL
 * @constructor
 */

const ROUTES = {
    LOGIN: '',
    REGISTRATION: 'registration',
    RECOVER_PASSWORD: 'reset-password',
    ALL_SOCIALS: 'other',
    SOCIALS_LOGIN: 'socials',
    USERNAME_LOGIN: 'username-login',
    ALL_SOCIALS_LOGIN: 'socials-login',
    SOCIALS: 'all-socials',
};

const IGNORELIST = [
    'onlyWidgets',
    'apiUrl',
    'defaultLoginUrl',
    'popupBackgroundColor',
    'iframeZIndex',
    'preloader',
    'widgetBaseUrl',
    'route',
    'inFullscreenMode',

    'redirectUrl',
    'widgetVersion',
    'projectId',

    'callbackUrl',
    'loginUrl',
    'state',
];

const DEFAULT_CONFIG = {
    apiUrl: 'https://login.xsolla.com/api/',
    onlyWidgets: false,
    defaultLoginUrl: 'https://xl-widget.xsolla.com/auth.html',
    popupBackgroundColor: 'rgb(187, 187, 187)',
    iframeZIndex: 1000000,
    preloader: '<div></div>',
    widgetBaseUrl: 'https://xl-widget.xsolla.com/',
    route: ROUTES.LOGIN,
    inFullscreenMode: false,
    response_type: 'code'
};

const INVALID_LOGIN_ERROR_CODE = 1;
const INCORRECT_LOGIN_OR_PASSWORD_ERROR_CODE = 2;

const IFRAME_ID = 'XsollaLoginWidgetIframe';
const widgetIframe = document.createElement('iframe');

class XL {
    constructor() {
        this.socialUrls = {};
        this.eventTypes = {
            LOAD: 'load',
            CLOSE: 'close',
            HIDE_POPUP: 'hide popup',
            REGISTRATION_REQUEST: 'registration request',
            AUTHENTICATED: 'authenticated'
        };

        // need for export purposes
        this.ROUTES = ROUTES;

        this.dispatcher = document.createElement('div');
        this.onHideEvent = this.onHideEvent.bind(this);
    }

    init(options) {
        this.config = Object.assign({}, DEFAULT_CONFIG, options);
        this.config.popupBackgroundColor = DEFAULT_CONFIG.popupBackgroundColor;
        this.api = new XLApi(options.projectId, this.config.apiUrl);

        const eventMethod = window.addEventListener ? 'addEventListener' : 'attachEvent';
        const eventer = window[eventMethod];
        const messageEvent = eventMethod === 'attachEvent' ? 'onmessage' : 'message';

        // Listen to message from child window
        eventer(messageEvent, (e) => {
            let event;
            if (typeof e.data === 'string') {
                // Old format - string only
                event = new CustomEvent(this.eventTypes[e.data]);
            } else {
                // New format - {type: 'event', ...}
                event = new CustomEvent(this.eventTypes[e.data.type], {detail: e.data});
            }
            this.dispatcher.dispatchEvent(event);
        }, false);

        Object.keys(this.eventTypes).map((eventKey) => {
            this.on(this.eventTypes[eventKey]);
        });

        if(options.popupBackgroundColor) {
            this.config.popupBackgroundColor = options.popupBackgroundColor;
        }

        this.dispatcher.addEventListener(this.eventTypes.HIDE_POPUP, this.onHideEvent);

        if (!this.config.onlyWidgets) {

            let params = {};
            params.projectId = options.projectId;
            if (this.config.redirectUrl) {
                params.redirect_url = this.config.redirectUrl;
            }
            if (this.config.loginUrl) {
                params.login_url = this.config.loginUrl;
            }
            if (this.config.callbackUrl) {
                params.login_url = this.config.callbackUrl;
            }
        }
    }

    /**
     * Performs login
     * @param prop
     * @param error - call in case error
     * @param success
     */
    login(prop, error, success) {

        if (!prop || !this.socialUrls) {
            return;
        }

        /**
         * props
         * authType: sn-<social name>, login-pass, sms
         */
        if (prop.authType) {
            if (prop.authType.startsWith('sn-')) {
                const socialUrl = this.socialUrls[prop.authType];
                if (socialUrl != undefined) {
                    window.location.href = this.socialUrls[prop.authType];
                } else {
                    console.error('Auth type: ' + prop.authType + ' doesn\'t exist');
                }

            } else if (prop.authType == 'login-pass') {
                this.api.loginPassAuth(prop.login, prop.pass, prop.rememberMe, this.config.redirectUrl, (res) => {
                    if (res.login_url) {
                        const finishAuth = function () {
                            window.location.href = res.login_url;
                        };
                        if (success) {
                            success({status: 'success', finish: finishAuth, redirectUrl: res.login_url});
                        } else {
                            finishAuth();
                        }
                    } else {
                        error(this.createErrorObject('Login or pass not valid', INCORRECT_LOGIN_OR_PASSWORD_ERROR_CODE));
                    }
                }, function (err) {
                    error(err);
                });
            } else if (prop.authType == 'sms') {
                if (smsAuthStep == 'phone') {
                    this.api.smsAuth(prop.phoneNumber, null, null);
                } else if (smsAuthStep == 'code') {

                }
            } else {
                console.error('Unknown auth type');
            }
        }
    }

    createErrorObject(message, code) {
        return {
            error: {
                message: message,
                code: code || -1
            }
        };
    };

    getProjectId() {
        return this.config.projectId;
    };

    getRedirectURL() {
        return this.config.redirectUrl;
    };

    getTheme() {
        return this.config.theme;
    }

    getCallbackUrl() {
        if (this.config.callbackUrl) {
            return this.config.callbackUrl;
        } else if (this.config.loginUrl) {
            return this.config.loginUrl;
        } else if (this.config.externalWindow) {
            return DEFAULT_CONFIG.defaultLoginUrl;
        }
    };

    /**
     * @deprecated use getLink instead
     */
    getIframeSrc() {
        return this.getLink.apply(this, arguments);
    }

    getLink(options = {}) {
        let widgetBaseUrl = options.widgetBaseUrl || this.config.widgetBaseUrl;

        if (widgetBaseUrl.substr(-1) !== '/') {
            widgetBaseUrl += '/';
        }

        const route = options.route || this.config.route;

        let src = widgetBaseUrl + route + '?projectId=' + encodeURIComponent(this.getProjectId()) + '&widget_sdk_version=' + version;

        let useOAuth2 = false;

        // Fields appended by loop
        // locale, fields, theme, compact, client_id, redirect_uri, response_type, state, externalWindow
        for (const option of Object.keys(this.config)) {
            if (!IGNORELIST.includes(option)) {
                const snakeOption = toSnakeCase(option);
                if (!useOAuth2 && snakeOption === 'client_id') {
                    useOAuth2 = true;
                }
                src += `&${snakeOption}=${encodeURIComponent(this.config[option])}`
            }
        }

        const redirectUrl = this.getRedirectURL();
        if (redirectUrl) {
            src = src + '&redirectUrl=' + encodeURIComponent(redirectUrl);
        }

        const callbackUrl = this.getCallbackUrl();

        if (callbackUrl) {
            src = src + '&login_url=' + encodeURIComponent(callbackUrl);
        }

        const {externalWindow, state} = this.config;
        if (externalWindow) {
            src = src + '&external_window=' + encodeURIComponent(externalWindow);
        }

        if (useOAuth2) {
            src += `&state=${encodeURIComponent(state || Math.random().toString(36).substring(2))}`
        }

        const widgetVersion = this.config.widgetVersion;
        if (widgetVersion) {
            src += '&version=' + encodeURIComponent(widgetVersion);
        }

        return src;
    }

    AuthWidget(elementId, options) {
        if (this.api) {
            if (!elementId) {
                console.error('No div name!');
            } else {
                if (options == undefined) {
                    options = {};
                }
                const width = `${options.width || 400}px`;
                const height = `${options.height || 550}px`;

                widgetIframe.onload = () => {
                    element.removeChild(preloader);
                    widgetIframe.style.width = '100%';
                    widgetIframe.style.height = '100%';
                    let event = new CustomEvent('load');
                    this.dispatcher.dispatchEvent(event);
                };
                widgetIframe.style.width = 0;
                widgetIframe.style.height = 0;
                widgetIframe.frameBorder = '0';
                widgetIframe.src = this.getIframeSrc(options);
                widgetIframe.id = IFRAME_ID;

                const preloader = document.createElement('div');

                preloader.innerHTML = this.config.preloader;

                const element = document.getElementById(elementId);
                if (element) {
                    element.style.width = width;
                    element.style.height = height;
                    element.appendChild(preloader);
                    element.appendChild(widgetIframe);
                } else {
                    console.error('Element \"' + elementId + '\" not found!');
                }

            }
        } else {
            console.error('Please run XL.init() first');
        }
    };

    onCloseEvent() {
        widgetIframe.parentNode.removeChild(widgetIframe);
    }

    _hide() {
        widgetIframe.style.position = '';
        widgetIframe.style.zIndex = '';
        widgetIframe.style.left = '';
        widgetIframe.style.top = '';
        widgetIframe.style.width = 0;
        widgetIframe.style.height = 0;
        widgetIframe.style.backgroundColor = '';
    }

    onHideEvent() {
        if (this.config.inFullscreenMode) {
            this._hide();
        }
    }

    /**
     * link event with handler
     * @param event
     * @param handler
     */

    on(event, handler) {
        handler = handler || function() {};

        if (event === this.eventTypes.CLOSE) {
            if (!handler) {
                handler = this.onCloseEvent;
            }
            else {
                this.dispatcher.removeEventListener(event, this.onCloseEvent);
            }
        }

        this.dispatcher.addEventListener(event, (e) => handler(e.detail));
    };

    _show() {
        widgetIframe.style.position = 'fixed';
        widgetIframe.style.zIndex = this.config.iframeZIndex;
        widgetIframe.style.left = '0';
        widgetIframe.style.top = '0';
        widgetIframe.style.width = '100%';
        widgetIframe.style.height = '100%';
        widgetIframe.style.backgroundColor = this.config.popupBackgroundColor;
        this.config.inFullscreenMode = true;
    }

    /**
     * open fullsreen popup for widget
     */

    show() {
        if (!document.getElementById(IFRAME_ID)) {
            widgetIframe.src = this.getIframeSrc();
            widgetIframe.id = IFRAME_ID;
            widgetIframe.style.width = 0;
            widgetIframe.style.height = 0;
            widgetIframe.frameBorder = '0';

            widgetIframe.onload = () => {
                let event = new CustomEvent('load');
                this.dispatcher.dispatchEvent(event);
            };
            this._show();

            document.body.appendChild(widgetIframe);
        } else {
            this._show();
        }
    };
}

const result = new XL();

module.exports = result;